File: 31.下一个排列.java
/*
 * @lc app=leetcode.cn id=31 lang=java
 *
 * [31] 下一个排列
 */

// @lc code=start
class Solution {
    public void nextPermutation(int[] nums) {
        int i = nums.length - 2;
        while (i >= 0 && nums[i] >= nums[i + 1])  --i;  // find the smaller digit
        if (i >= 0) {
            int j = nums.length - 1;
            while (j >= 0 && nums[i] >= nums[j])  --j;  // find the larger digit
            swap(nums, i, j);   // swap the digits
        }
        reverse(nums, i + 1);   // ensure nums[i+1:n] is ascending
    }

    private static void swap(int[] nums, int i, int j) {
        int tmp = nums[j];
        nums[j] = nums[i];
        nums[i] = tmp;
    }

    private static void reverse(int[] nums, int begin) {
        int l = begin, r = nums.length - 1;
        while (l < r) {
            swap(nums, l, r);
            ++l;
            --r;
        }
    }
}
// @lc code=end

File: 31.下一个排列.py
#
# @lc app=leetcode.cn id=31 lang=python3
#
# [31] 下一个排列
#
from typing import List

# @lc code=start
class Solution:
    def nextPermutation(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        n = len(nums)
        if n < 2:
            return
        idx1 = 0
        for i in range(n-1, 0, -1):
            if nums[i] > nums[i-1]:
                idx1 = i
                break
        if idx1 == n-1:
            nums[idx1], nums[idx1-1] = nums[idx1-1], nums[idx1]
        else:
            self.reverse_inplace(nums, idx1, n-1)   # ensure the next permutation, not just a greater one
            if not idx1:
                return
            for idx2 in range(idx1, n):
                if nums[idx2] > nums[idx1-1]:
                    nums[idx1-1], nums[idx2] = nums[idx2], nums[idx1-1]
                    break
    
    def reverse_inplace(self, nums, start, end):
        mid = (start + end + 1) >> 1
        for k, j in enumerate(range(start, mid)):
            nums[j], nums[end-k] = nums[end-k], nums[j]


# @lc code=end
# s = Solution()
# l = [1, 3, 2]
# s.nextPermutation(l)
# print(l)
